packages/labs/react/src/create-component.ts

```diff
@@ -185,215 +189,121 @@ const setRef = (ref: React.Ref<unknown>, value: Element | null) => {
-          // coerce it to `class` so it's handled correctly.
-          props[k === 'className' ? 'class' : k] = v;
-          continue;
+    // useLayoutEffect produces warnings during server rendering.
+    if (!NODE_MODE) {
+      // This one has no dependency array so it'll run on every re-render.
+      React.useLayoutEffect(() => {
+        if (elementRef.current === null) {
+          return;
         }
-
-        if (eventProps.has(k) || k in element.prototype) {
-          this._elementProps[k] = v;
-          continue;
+        for (const prop in elementProps) {
+          setProperty(
+            elementRef.current,
+            prop,
+            props[prop],
+            prevPropsRef.current ? prevPropsRef.current[prop] : undefined,
+            events
+          );
         }
+        // Note, the spirit of React might be to "unset" any old values that
+        // are no longer included; however, there's no reasonable value to set
+        // them to so we just leave the previous state as is.
 
-        props[k] = v;
-      }
+        prevPropsRef.current = props;
+      });
+
+      // Empty dependency array so this will only run once after first render.
+      React.useLayoutEffect(() => {
+        elementRef.current?.removeAttribute('defer-hydration');
+      }, []);
+    }
 
-      // If component is to be server rendered with `@lit-labs/ssr-react` pass
+    if (NODE_MODE) {
+      // If component is to be server rendered with `@lit-labs/ssr-react`, pass
       // element properties in a special bag to be set by the server-side
       // element renderer.
       if (
-        NODE_MODE &&
-        createElement.name === 'litPatchedCreateElement' &&
-        Object.keys(this._elementProps).length
+        React.createElement.name === 'litPatchedCreateElement' &&
+        Object.keys(elementProps).length
       ) {
         // This property needs to remain unminified.
-        props['_$litProps$'] = this._elementProps;
-        return createElement<React.HTMLAttributes<I>, I>(tag, props);
+        reactProps['_$litProps$'] = elementProps;
       }
-
-      if (!NODE_MODE) {
-        // Suppress hydration warning for server-rendered attributes, including
-        // "defer-hydration"
-        props['suppressHydrationWarning'] = true;
-      }
-
-      return createElement<React.HTMLAttributes<I>, I>(tag, props);
+    } else {
+      // Suppress hydration warning for server-rendered attributes.
+      // This property needs to remain unminified.
+      reactProps['suppressHydrationWarning'] = true;
     }
-  }
 
-  const ForwardedComponent: ReactWebComponent<I, E> = ReactModule.forwardRef<
-    I,
-    WebComponentProps<I, E>
-  >((props, __forwardedRef) =>
-    createElement<Props, ReactComponent, typeof ReactComponent>(
-      ReactComponent,
-      {...props, __forwardedRef}
-    )
-  );
+    return React.createElement(tagName, {
+      ...reactProps,
+      ref: (node: I) => {
```

Comment:
This creates a new closure every render. Can we hoist it?

Reply:
It needs to be declared in the scope of the function component so we can assign to `elementRef.current` as well. I think a new closure every render here is unavoidable.